### PR TITLE
Fix labels synchronization issues on Cilium

### PR DESCRIPTION
**Pull request by aanm**
Read on a per commit basis.

```release-note
Fixed label synchronization issues in Cilium, ensuring accurate representation of endpoint labels during restoration and addressing out-of-sync problems caused by label changes while the Cilium agent is down.
```

### CODE REVIEW

File: pkg/labels/oplabels.go
Comment (gandro) [inline pkg/labels/oplabels.go:121]: Note for future reviewers: During review, I was confused why this wasn't touching the `upsertLabels` logic, as the commit description clearly talks about replacement of labels, not just about "keeping". This is addressed in the next commit.

File: pkg/labels/oplabels.go
Comment (gandro) [inline pkg/labels/oplabels.go:219]: Shouldn't we pass in the `source` from `Replace{Identity,Info}Labels` and compare to that? I find it a bit weird that

- for upsert, we only compare sources to existing labels, but for anything new we don't consider the source
- for delete (aka keep), we consider the the source passed in the previous commit

In other words (assuming an empty endpoint that doesn't have any labels), I would expect that if you pass in `labels={cilium:foo=bar,nginx:baz=qux},source=nginx`  to `Replace{Identity,Info}Labels`, that it would only insert the `nginx` label, and ignore the "cilium:foo" label. 

But currently, if there is no `foo` label, it will also insert the `cilium:foo` label.

Is this behavior intentional? If so, why?

File: pkg/endpoint/endpoint.go
Comment (gandro) [inline pkg/endpoint/endpoint.go]: Why not use `e.K8sNamespaceAndPodNameIsSet()`? We already have this check at the call site, why is this needed here?

https://github.com/cilium/cilium/blob/6feffd1ea1c091fc33b96ac80e50bb0ec570d1c0/daemon/cmd/endpoint.go#L467

File: pkg/endpoint/endpoint.go
Comment (gandro) [inline pkg/endpoint/endpoint.go]: godoc comment says `If this list is empty or nil` but we only check for `nil` here. Any reason we're not using `len(initialLabels) != 0` here?

File: pkg/endpoint/endpoint.go
Comment (gandro) [inline pkg/endpoint/endpoint.go]: So if the users sets `initialLabels`, then we're deleting any previously set labels not present in set of `initialLabels +  resolveMetadata()` . I think this needs be documented in the go docs, as currently it only talks about the case where `initialLabels` is unset (which is the common case).

**Review by gandro — CHANGES_REQUESTED** [review]
Thanks! Looks good overall, but I have some questions around the details

File: pkg/labels/oplabels.go
Comment (aanm) [reply pkg/labels/oplabels.go:219]: That's a fair question. I also thought about this.

1. The issue was that if the "source" is defined in the label, having another parameter to receive the "source" would look odd?
2. Remove the "source" parameter from the 'deleteUnMarked' and move the logic of "what to keep" outside of this function.

Should I go with 2 and keep this current change?

File: pkg/endpoint/endpoint.go
Comment (aanm) [reply pkg/endpoint/endpoint.go]: ah good point, will do.

**Review by nebril — APPROVED** [review]
LGTM aside from Sebastian's comments.

File: pkg/endpoint/endpoint.go
Comment (aanm) [reply pkg/endpoint/endpoint.go]: Yes, during endpoint creation, the `initialLabels` are initialized which means that we want to set the endpoint with the init identity. I've reworded the godoc with the following:
```
// If this list is empty, the labels set on the endpoint will be
// replaced by the labels returned 'MetadataResolverCB' by as long their source
// matches the original source.
// If this list is nil and the resolver is unable to retrieve the endpoint
// labels from k8s, the endpoint will be set with the 'init' identity.
```

File: pkg/endpoint/endpoint.go
Comment (aanm) [reply pkg/endpoint/endpoint.go]: Yes, does this look good?
```
// If this list is empty, the labels set on the endpoint will be
// replaced by the labels returned 'MetadataResolverCB' by as long their source
// matches the original source.
// If this list is nil and the resolver is unable to retrieve the endpoint
// labels from k8s, the endpoint will be set with the 'init' identity.
```

File: pkg/endpoint/endpoint.go
Comment (gandro) [reply pkg/endpoint/endpoint.go]: I see. Honestly, I find this a bit fragile that `nil` and `empty list` behave differently. Have you considered just introducing a boolean for the init case?

File: pkg/labels/oplabels.go
Comment (gandro) [reply pkg/labels/oplabels.go:219]: > The issue was that if the "source" is defined in the label, having another parameter to receive the "source" would look odd?

To me, this doesn't seem particularly odd. `source` acts like a filter on labels, we could even rename it to `sourceFilter` to make it more clear. If you provide a source, then it becomes clear that anything within `labels` which does not match the `sourceFilter` will be ignored.

> Should I go with 2 and keep this current change?

The parameter would still exist for `Replace{Identity,Info}Labels`, right? In that case it doesn't make a difference to me, since my concern is more with the (imo surprising) behavior and less with the function signatures.

File: pkg/endpoint/endpoint.go
Comment (gandro) [reply pkg/endpoint/endpoint.go]: The new doc comment seems to address the init case above, but it does not mention the behavior I pointed out here, namely that we delete any existing labels on the endpoint if the list is _not_ empty.

**Review by learnitall — APPROVED** [review]
LGTM pending @gandro's approval

**Comment by gandro** [comment]
@aanm Why did you ask for a re-review without any changes to the code?

File: pkg/labels/oplabels.go
Comment (tklauser) [reply pkg/labels/oplabels.go:219]: > To me, this doesn't seem particularly odd. `source` acts like a filter on labels, we could even rename it to `sourceFilter` to make it more clear. If you provide a source, then it becomes clear that anything within `labels` which does not match the `sourceFilter` will be ignored.

That's the semantics I assumed during my review before I noticed you already commented here. I agree that renaming the parameter to `sourceFilter` and documenting it in the godoc would make its purpose clearer.

File: daemon/cmd/endpoint.go
Comment (tklauser) [inline daemon/cmd/endpoint.go]: nit: let's `s/infoLbls/k8sInfoLbls` to make the `MergeLabels` call below easier to read

```suggestion
		pod, cp, k8sIdentityLbls, k8sInfoLbls, annotations, err := d.fetchK8sMetadataForEndpoint(ep.K8sNamespace, ep.K8sPodName)
```

File: daemon/cmd/endpoint.go
Comment (tklauser) [inline daemon/cmd/endpoint.go]: per the comment above:

```suggestion
			infoLabels.MergeLabels(k8sInfoLbls)
```

**Review by tklauser — COMMENTED** [review]
One small naming nit. Otherwise LGTM once Sebastian's review comments are addressed.

File: pkg/endpoint/endpoint.go
Comment (joestringer) [inline pkg/endpoint/endpoint.go]: Strictly speaking this should probably also receive the real label source from the callers (eg k8s Pod updates or Cilium API) to avoid overwriting labels from other sources(?).

File: pkg/labels/oplabels.go
Comment (gandro) [inline pkg/labels/oplabels.go:206]: > The new label must also match the old label 'source' in order for it to be replaced.

nit (non-blocking): I was slightly confused why the commit adding this godoc comment did not also add the check (the check is only added in the next commit). Might be worth moving that part of the go doc change into the next commit.

File: pkg/endpoint/endpoint.go
Comment (gandro) [inline pkg/endpoint/endpoint.go]: I'm still not super happy with this go doc comment. initialLabels will **replace** cause any existing labels to be replaced with the set of `initialLabels +  resolveMetadata()`. The fact that is says "added" sounds like it's only additive, but it's not, it's a replacement.

File: pkg/endpoint/endpoint.go
Comment (gandro) [inline pkg/endpoint/endpoint.go]: nit: Document the meaning of the return value

File: pkg/endpoint/endpoint.go
Comment (gandro) [inline pkg/endpoint/endpoint.go]: The `During restore the 'initialLabels' is nil` comment might be still correct, but it's now irrelevant here. I'd remove it.

File: pkg/endpoint/endpoint.go
Comment (gandro) [inline pkg/endpoint/endpoint.go:1682]: Any reason we don't log the `Unable to fetch kubernetes labels` error here?

File: pkg/endpoint/endpoint.go
Comment (gandro) [inline pkg/endpoint/endpoint.go]: Do we have a long-term plan to properly fix this? For example using jobs?

**Review by gandro — COMMENTED** [review]
Logic-wise, this looks mostly fine to me now. Mainly nits regarding source code comments that in my opinion don't match what the code is doing

File: pkg/endpoint/endpoint.go
Comment (gandro) [inline pkg/endpoint/endpoint.go]: In terms of controller-lifetime, this looks good to me now. But this raises the question:

Let's assume `resolve-endpoint-1-xwing` is called with `baseLabels={foo:bar1}`, but it fails a few times (e.g. because api-server is overloaded and resolveMetadata returns errors) and is in a back-off period for a couple of seconds.

Meanwhile, api-server recovers, `resolve-endpoint-2-xwing` is called with `baseLabels={foo:bar2}`, and immediately succeeds. Then, the  `resolve-endpoint-1-xwing` is woken up again, also suceeds, but now overwrites the labels with the outdated `baseLabels={foo:bar1}`.

Is that a problem? Should we somehow track the resolverIndex inside `DoFunc` and give up if we see that we got replaced by a different controller?

**Review by gandro — APPROVED** [review]
Code-wise looks good to me. I have one more question around the controller lifetime, which I don't think it can be hit with the current callers to `RunMetadataResolver` (as they don't provide baseLabels), but might be in the future.

File: pkg/endpoint/endpoint.go
Comment (joestringer) [inline pkg/endpoint/endpoint.go]: Why is this passing `labels.LabelSourceAny`?

This function doesn't strictly know that the source is `LabelSourceK8s` because the `resolveMetadata` callback parameter is what injects the k8s metadata resolver. That said, I see a few possible options here:
- `RunMetadataResolver` could receive a source parameter so that the source could be aligned with the actual source of the labels, or
- `MetadataResolverCB` could return the source of the metadata, or
- This could just take `labels.LabelSourceK8s` since all the callers feed the labels/metadata from k8s anyway.

File: pkg/endpoint/endpoint.go
Comment (joestringer) [inline pkg/endpoint/endpoint.go:1772]: nit: I feel like the refactor here could be split from the commit `pkg/endpoint: keep endpoint labels for their original sources`; The >200 line size of the commit with two potential fixes in there makes it harder to track at a glance whether there are important changes being made as part of this refactor within the bugfix patch, or this is a pure refactor to move the content that was previously directly here in the `DoFunc` declaration into its own function.

Upon a closer look, this *is* making around 150LoC refactor as part of a patch that changes behaviour. :(

File: pkg/endpoint/endpoint.go
Comment (joestringer) [inline pkg/endpoint/endpoint.go]: I feel like the `len(baseLabels) != 0` logic contradicts the comment. If there's base labels, then k8s labels can also overwrite the base labels. Then we reintroduce one of the bugs that this patch aims to solve?

Should this be doing two calls or passing down two sets of labels where the base labels use `LabelSourceAny` (and potentially happen after / override the k8s labels?)?

File: pkg/endpoint/endpoint.go
Comment (joestringer) [reply pkg/endpoint/endpoint.go]: I can't help but feel like `pkg/controller` should just have an option to "Remove controller once complete" so that we don't have this code in `pkg/endpoint` trying to make sure that multiple separate calls into this function don't go and delete other references of the controller with the same name. Creating multiple controllers and tracking these indexes and so on just seems more complicated than we need to ensure that there's only one copy of a controller at once and that it is removed dynamically once complete.

Incidentally, now that this controller doesn't have a `RunInterval`, do we even need to remove it after completion at all?

**Review by tklauser — APPROVED** [review]
LGTM once Joe's unresolved review comments are addressed.

File: pkg/endpoint/endpoint.go
Comment (aanm) [reply pkg/endpoint/endpoint.go]: I didn't do it to avoid changing more code but does make sense! :bulb:

File: pkg/endpoint/endpoint.go
Comment (aanm) [reply pkg/endpoint/endpoint.go]: This is on the error case, where we want to replace all labels because we want to put the endpoint in "init" mode. Kind of unrelated but I realized that `info` should not be passed and it should be `nil` for this case.

File: pkg/endpoint/endpoint.go
Comment (aanm) [reply pkg/endpoint/endpoint.go]: That's actually way simpler!

File: pkg/endpoint/endpoint.go
Comment (aanm) [reply pkg/endpoint/endpoint.go]: The comment is indeed confusing, is this better?
> 	// If 'baseLabels' are not set then 'controllerBaseLabels' only contains
>	// labels from k8s. Thus, we should only replace the labels that have their
>	// source as 'k8s' otherwise we will risk on replacing other labels that
>	// were added from other sources.

File: pkg/endpoint/endpoint.go
Comment (aanm) [reply pkg/endpoint/endpoint.go:1772]: oh yeah. Sorry about that. I tried to reason with doing the refactor on a separate commit but I gave up because it was kind of fixing the bug. IIRC this code didn't exist in the first review which made it even further difficult to split it in seperate commits.

File: pkg/endpoint/endpoint.go
Comment (joestringer) [reply pkg/endpoint/endpoint.go]: Ah right, that makes sense.

File: pkg/endpoint/endpoint.go
Comment (joestringer) [reply pkg/endpoint/endpoint.go]: That context helps, but I think we still have an issue?

If `baseLabels` *is* set, then any labels we learn from Kubernetes could overwrite labels from other sources.

File: pkg/endpoint/endpoint.go
Comment (joestringer) [reply pkg/endpoint/endpoint.go:1772]: Refactoring the function in a first patch like this can drop ~100 lines already, then the functional changes will be smaller on top:

```
diff --git a/pkg/endpoint/endpoint.go b/pkg/endpoint/endpoint.go
index e6db5802552b..36176f3a2c96 100644
--- a/pkg/endpoint/endpoint.go
+++ b/pkg/endpoint/endpoint.go
@@ -1630,6 +1630,53 @@ func (e *Endpoint) APICanModifyConfig(n models.ConfigurationMap) error {
        return nil
 }

+// metadataResolver will resolve the endpoint's metadata from a metadata
+// resolver.
+//
+//   - bwm - the bandwidth manager used to update the bandwidth policy for this
+//     endpoint.
+//
+//   - resolveMetadata - the metadata resolver that will be used to retrieve this
+//     endpoint's metadata.
+func (e *Endpoint) metadataResolver(ctx context.Context,
+       bwm bandwidth.Manager,
+       resolveMetadata MetadataResolverCB) (err error) {
+       ns, podName := e.GetK8sNamespace(), e.GetK8sPodName()
+       pod, cp, identityLabels, info, _, err := resolveMetadata(ns, podName)
+       if err != nil {
+               e.Logger(resolveLabels).WithError(err).Warning("Unable to fetch kubernetes labels")
+               return err
+       }
+       e.SetPod(pod)
+       e.SetK8sMetadata(cp)
+       e.UpdateNoTrackRules(func(_, _ string) (noTrackPort string, err error) {
+               po, _, _, _, _, err := resolveMetadata(ns, podName)
+               if err != nil {
+                       return "", err
+               }
+               value, _ := annotation.Get(po, annotation.NoTrack, annotation.NoTrackAlias)
+               return value, nil
+       })
+       e.UpdateVisibilityPolicy(func(_, _ string) (proxyVisibility string, err error) {
+               po, _, _, _, _, err := resolveMetadata(ns, podName)
+               if err != nil {
+                       return "", err
+               }
+               value, _ := annotation.Get(po, annotation.ProxyVisibility, annotation.ProxyVisibilityAlias)
+               return value, nil
+       })
+       e.UpdateBandwidthPolicy(bwm, func(ns, podName string) (bandwidthEgress string, err error) {
+               _, _, _, _, annotations, err := resolveMetadata(ns, podName)
+               if err != nil {
+                       return "", err
+               }
+               return annotations[bandwidth.EgressBandwidth], nil
+       })
+       e.UpdateLabels(ctx, identityLabels, info, true)
+
+       return nil
+}
+
 // MetadataResolverCB provides an implementation for resolving the endpoint
 // metadata for an endpoint such as the associated labels and annotations.
 type MetadataResolverCB func(ns, podName string) (pod *slim_corev1.Pod, _ []slim_corev1.ContainerPort, identityLabels labels.Labels, infoLabels labels.Labels, annotations map[string]string, err error)
@@ -1658,38 +1705,10 @@ func (e *Endpoint) RunMetadataResolver(bwm bandwidth.Manager, resolveMetadata Me
                controller.ControllerParams{
                        Group: resolveLabelsControllerGroup,
                        DoFunc: func(ctx context.Context) error {
-                               ns, podName := e.GetK8sNamespace(), e.GetK8sPodName()
-                               pod, cp, identityLabels, info, _, err := resolveMetadata(ns, podName)
+                               err := e.metadataResolver(ctx, bwm, resolveMetadata)
                                if err != nil {
-                                       e.Logger(resolveLabels).WithError(err).Warning("Unable to fetch kubernetes labels")
                                        return err
                                }
-                               e.SetPod(pod)
-                               e.SetK8sMetadata(cp)
-                               e.UpdateNoTrackRules(func(_, _ string) (noTrackPort string, err error) {
-                                       po, _, _, _, _, err := resolveMetadata(ns, podName)
-                                       if err != nil {
-                                               return "", err
-                                       }
-                                       value, _ := annotation.Get(po, annotation.NoTrack, annotation.NoTrackAlias)
-                                       return value, nil
-                               })
-                               e.UpdateVisibilityPolicy(func(_, _ string) (proxyVisibility string, err error) {
-                                       po, _, _, _, _, err := resolveMetadata(ns, podName)
-                                       if err != nil {
-                                               return "", err
-                                       }
-                                       value, _ := annotation.Get(po, annotation.ProxyVisibility, annotation.ProxyVisibilityAlias)
-                                       return value, nil
-                               })
-                               e.UpdateBandwidthPolicy(bwm, func(ns, podName string) (bandwidthEgress string, err error) {
-                                       _, _, _, _, annotations, err := resolveMetadata(ns, podName)
-                                       if err != nil {
-                                               return "", err
-                                       }
-                                       return annotations[bandwidth.EgressBandwidth], nil
-                               })
-                               e.UpdateLabels(ctx, identityLabels, info, true)
                                close(done)
                                return nil
                        },
```

Comment (joestringer) [reply pkg/endpoint/endpoint.go:1772]: For review purposes I just rebased the PR on top of this incremental patch, and it makes that core fix change much more targeted to the changes that are being introduced: https://github.com/joestringer/cilium/tree/pr/fix-cni-add

File: pkg/endpoint/endpoint.go
Comment (joestringer) [reply pkg/endpoint/endpoint.go]: Nice, a solution where we basically just delete code :100:

File: pkg/endpoint/endpoint.go
Comment (joestringer) [reply pkg/endpoint/endpoint.go]: New version LGTM

File: pkg/endpoint/endpoint.go
Comment (joestringer) [reply pkg/endpoint/endpoint.go:1772]: This made it a lot easier to read the changes and check that they all line up. I don't see anything else in this patch after splitting it out. There's still some minor things I'd have a mild preference to further split, but there's few enough changes in a single patch that I can keep it all in my head during review.

File: pkg/endpoint/endpoint.go
Comment (joestringer) [reply pkg/endpoint/endpoint.go]: Ah, I think that implicitly if `baseLabels` is set then the endpoint is being created, so updating the identity cannot overwrite labels, because there are no labels to overwrite(?)

In that case I think it's logically correct, just confusing.

Maybe a particularly weird case would be, what if the API labels during endpoint create include a label with source k8s? Will Cilium sanitize that input, could it conflict with k8s labels?

File: pkg/endpoint/endpoint.go
Comment (joestringer) [reply pkg/endpoint/endpoint.go:1772]: Feel free to take these or leave them. I've reviewed the rest of the patch at this point.

File: pkg/endpoint/endpoint.go
Comment (aanm) [reply pkg/endpoint/endpoint.go]: > Ah, I think that implicitly if `baseLabels` is set then the endpoint is being created, so updating the identity cannot overwrite labels, because there are no labels to overwrite(?)

Correct.
> 
> In that case I think it's logically correct, just confusing.
> 
> Maybe a particularly weird case would be, what if the API labels during endpoint create include a label with source k8s? Will Cilium sanitize that input, could it conflict with k8s labels?

If the endpoint create contains let's say "k8s:foo=bar" and then the metadataResolver, which has the k8s source, fetches the data with "foo=baz" then the final labels will be "k8s:foo=baz". Just because they have the same source.

If we had "cni:foo=bar", then the final labels would be "cni:foo=bar". The k8s won't replace the initial labels.

Comment (joestringer) [reply pkg/endpoint/endpoint.go]: I see. That sounds like the right behaviour.

If the underlying `UpdateLabels` implementation provides that guarantee then I still don't understand the need for this if conditional, because labels from k8s should only ever have the k8s source, so they should never overwrite labels from other sources. Is this just defensive coding in case something changes in future?

I looked around but I couldn't find a specific test that validated the difference in behaviour for this conditional based on the input labels.

Comment (joestringer) [reply pkg/endpoint/endpoint.go]: Answer: The `UpdateLabels` API in the endpoint is quirky in a few ways that makes this specific pattern necessary:
- If you use `UpdateLabels(..., SourceAny, newLabels, ...)`, it is equivalent to deleting all labels in the endpoint and adding all labels in the `newLabels` set.
- If you use `UpdateLabels(..., SourceK8s, newLabels, ...)`, it is equivalent to deleting all labels in the endpoint with k8s source and adding the labels in the `newLabels` set.

For endpoint creation, we know that some labels may come from the API, and we know that the endpoint has no labels currently, so we can replace all labels. After endpoint creation (including during restore), resolving the metadata from k8s will only present new k8s labels, and will *not* retrieve the source of truth for non-k8s labels. Therefore, if we use `SourceAny` with the metadata resolver, it could end up deleting all non-k8s labels.

I think that the `UpdateLabels` API here is overly complex and is what causes some of these complications, but this code is actually correct and necessary as-is.

As follow-ups I think it would be nice to rethink how the Endpoint's label-setting APIs work to make them less confusing.

**Pull request merged**
